modules/assim.sequential/R/downscale_function.R

```diff
@@ -1,88 +1,284 @@
+##' @param date In quotes, if SDA site run, format is yyyy/mm/dd, if NEON, yyyy-mm-dd. Restricted to years within file supplied to 'data'.
+##' @param carbon_pool In quotes, carbon pool of interest. Name must match carbon pool name found within file supplied to 'data'.
+##' @param covariates SpatRaster stack, used as predictors in CNN. Layers within stack should be named. Recommended that this stack be generated using 'covariates' instructions in assim.sequential/inst folder
+##' @param model_type Character. Either "rf" for Random Forest or "cnn" for Convolutional Neural Network.
+##' @details This function will downscale forecast data to unmodeled locations using covariates and site locations
+##'
+##' @description This function uses either Random Forest or Convolutional Neural Network model based on the model_type parameter.
+##'
+##' @return It returns the `downscale_output` list containing lists for the training and testing data sets, models, and predicted maps for each ensemble member.
+
+
+##' @title North America Downscale Function
+##' @name SDA_downscale
+##' @author Joshua Ploshay, Sambhav Dixit
+##'
+##' @param preprocessed List. Preprocessed data returned as an output from the SDA_downscale_preprocess function.
+##' @param date Character. If SDA site run, format is yyyy/mm/dd; if NEON, yyyy-mm-dd. Restricted to years within file supplied to 'data'.
+##' @param carbon_pool Character. Carbon pool of interest. Name must match carbon pool name found within file supplied to 'data'.
+##' @param covariates SpatRaster stack. Used as predictors in CNN. Layers within stack should be named.
+##' @param model_type Character. Either "rf" for Random Forest or "cnn" for Convolutional Neural Network.
+##' @details This function will downscale forecast data to unmodeled locations using covariates and site locations
+##'
+##' @description This function uses either Random Forest or Convolutional Neural Network model based on the model_type parameter.
+##'
+##' @return A list containing the training and testing data sets, models, predicted maps for each ensemble member, and predictions for testing data.
+
+SDA_downscale <- function(preprocessed, date, carbon_pool, covariates, model_type) {
+  input_data <- preprocessed$input_data
+  site_coordinates <- preprocessed$site_coordinates
+  carbon_data <- preprocessed$carbon_data
+  
+  # Convert site coordinates to SpatVector
+  site_coordinates <- terra::vect(site_coordinates, geom = c("lon", "lat"), crs = "EPSG:4326")
+  
+  # Extract predictors from covariates raster using site coordinates
+  predictors <- as.data.frame(terra::extract(covariates, site_coordinates, ID = FALSE))
+  
+  # Dynamically get covariate names
+  covariate_names <- names(predictors)
+  
+  # Create a single data frame with all predictors and ensemble data
+  full_data <- cbind(carbon_data, predictors)
+  
+  # Split the observations into training and testing sets
+  set.seed(123)  # for reproducibility
```

Comment:
I know this is a legacy from Josh's code, but this could have unintended consequences and so should be addressed in future PRs, perhaps by making the seed an optional function argument (e.g. with a NULL default) and only running set.seed if it's provided

Reply:
issue addressed and solved in [4d2c6a5](https://github.com/PecanProject/pecan/pull/3321/commits/4d2c6a54db51315b872c9d3ba2ce1b527e459df4) commit .